sync/threadgroup.go

```diff
@@ -58,16 +66,33 @@ func (tg *ThreadGroup) Done() {
 	tg.wg.Done()
 }
 
-// Stop closes the Threadgroup's stopChan and blocks until the counter is
-// zero.
+// Stop closes the ThreadGroup's stopChan, closes members of the closer set,
+// and blocks until the counter is zero.
 func (tg *ThreadGroup) Stop() error {
 	tg.mu.Lock()
 	if tg.IsStopped() {
 		tg.mu.Unlock()
 		return ErrStopped
 	}
 	close(tg.stopChan)
+	for _, c := range tg.closers {
+		c.Close()
+	}
+	tg.closers = nil
 	tg.mu.Unlock()
 	tg.wg.Wait()
 	return nil
 }
+
+// RegisterCloser adds an io.Closer to the ThreadGroups closer set. Members of
+// the set will be closed when Stop is called. Note that this means the errors
+// returned by Close are not recoverable.
```

Comment:
Also note that this method should be exclusively called for threads which will be alive until `Close` is called. Good for `net.Listener`, not good for `net.Conn`. Also good for things like the renter repair loop and the hostdb scan loop.